File: payment-client/src/app/utilities/bcd-converter.ts
export class BcdConverter {
  static stringToBcd(input: string): string {
    if (input.length % 2 !== 0) {
      input = '0' + input;
    }
    let bcd = '';
    for (let i = 0; i < input.length; i += 2) {
      bcd += String.fromCharCode(parseInt(input.substr(i, 2), 16));
    }
    return bcd;
  }

  static bcdToString(bcd: string): string {
    let result = '';
    for (let i = 0; i < bcd.length; i++) {
      const byte = bcd.charCodeAt(i);
      result += ('0' + (byte >> 4).toString(16)).slice(-1);
      result += ('0' + (byte & 0x0f).toString(16)).slice(-1);
    }
    return result.replace(/^0+/, ''); // Remove leading zeros
  }
}

File: PaymentGateway/PaymentGateway.Application/Utilities/BcdConverter.cs
using System.Text;

namespace PaymentGateway.Application.Utilities
{
    public static class BcdConverter
    {
        public static byte[] StringToBcd(string input)
        {
            if (input.Length % 2 != 0)
            {
                input = "0" + input;
            }

            byte[] bcd = new byte[input.Length / 2];
            for (int i = 0; i < bcd.Length; i++)
            {
                bcd[i] = byte.Parse(input.Substring(2 * i, 2), System.Globalization.NumberStyles.HexNumber);
            }
            return bcd;
        }

        public static string BcdToString(byte[] bcd)
        {
            StringBuilder sb = new StringBuilder(bcd.Length * 2);
            foreach (byte b in bcd)
            {
                sb.Append(b.ToString("X2"));
            }
            return sb.ToString().TrimStart('0'); // Remove leading zeros
        }
    }
}
